### ISSUE TITLE
Version Tracking on two copies of the same plain-jane Visual Studio 2022 "Hello World" Windows C++ Console App causes MemoryAccessException upon Automatic Version Tracking.

### ISSUE THREAD

**Issue by AshleyT3** [NationalSecurityAgency/ghidra#6238]
**Describe the bug**
Version Tracking on two copies of the same plain-jane Visual Studio 2022 "Hello World" Windows C++ Console App causes MemoryAccessException upon Automatic Version Tracking.

**To Reproduce**
1. Install/start Visual Studio 2022 Community Edition.
2. Create New Project, Windows C++ Console App, click Next, project name is SimpleConsoleApp, select Location, check "Place solution... in the same directory", click Create.
3. With boilerplate comments removed, the following default Hello World should remain as follows...
```
#include <iostream>
int main()
{
    std::cout << "Hello World!\n";
}
```
4. With Debug/x64 selected, Build/Build Solution.
5. Right-click SimpleConsoleApp project, choose Open Folder in File Explorer, navigate to .\SimpleConsoleApp\x64\Debug where the .exe and .pdb are located.
6. Select both .exe/.pdb, use Copy/Paste to duplicate them (i.e., select both, Ctrl+C, Ctrl+V), then rename the copies to be SimpleConsoleApp2.exe and SimpleConsoleApp2.pdb. All these files should be in the same .\Debug output folder. These represent two binaries which are equal. We will be create a Ghidra Version Tracking session using SimpleConsoleApp.exe and SimpleConsoleApp2.exe.
7. Start Ghidra.
8. File/New Project, Non-Shared Project, Project Name is SimpleConsoleAppVtProj, click Finish.
9. Press I to import, locate and select SimpleConsoleApp.exe, click Select File To Import, click OK on the next Import dialog, click OK on the Import Results Summary dialog.
10. Repeat step 9 but for SimpleConsoleApp2.exe.
11. Highlight both SimpleConsoleApp.exe and SimpleConsoleApp2.exe, right-click and choose Open With, Version Tracking, SimpleConsoleApp.exe is the Source Program, SimpleConsoleApp2.exe is the Destination Program, click Next, click Run Precondition Checks, click Next, click Finish...
12. Do _not_ choose to auto-analyze, instead click the dialog box X or No to skip auto-analysis for both binaries. This is done simply to go incrementally, where it is desired to manually locate/load the .pdb first then auto-analyze, not that this is required but these are the steps taken. It also confirms symbols were loaded properly fwiw.
13. Ghidra VT will open both binaries in analysis windows... in the SimpleConsoleApp.exe window's menu choose Edit/Symbol Server Config, click +, Directory, specify the path to the SimpleConsoleApp.exe pdb file, click OK, click OK.
14. File/Load PDB File..., click Load, it should load, click OK to dismiss warnings. 
15. Press A to auto-analyze SimpleConsoleApp.exe.
16. Repeat steps 14 and 15 for SimpleConsoleApp2.exe.
17. Choose to Save All for both analysis windows.
18. Go to Version Tracking Window, File/Automatic Version Tracking, observe the following exception...
```
ghidra.program.model.mem.MemoryAccessException: Couldn't get all bytes for CodeUnit
java.lang.RuntimeException: ghidra.program.model.mem.MemoryAccessException: Couldn't get all bytes for CodeUnit
	at ghidra.app.plugin.match.MatchData.extractSourceHashes(MatchData.java:83)
	at ghidra.app.plugin.match.MatchData.matchData(MatchData.java:47)
	at ghidra.feature.vt.api.correlator.program.DataMatchProgramCorrelator.doCorrelate(DataMatchProgramCorrelator.java:61)
	at ghidra.feature.vt.api.util.VTAbstractProgramCorrelator.correlate(VTAbstractProgramCorrelator.java:73)
	at ghidra.feature.vt.gui.actions.AutoVersionTrackingTask.correlateAndPossiblyApply(AutoVersionTrackingTask.java:587)
	at ghidra.feature.vt.gui.actions.AutoVersionTrackingTask.doRun(AutoVersionTrackingTask.java:193)
	at ghidra.feature.vt.gui.actions.AutoVersionTrackingTask.run(AutoVersionTrackingTask.java:123)
	at ghidra.util.task.Task.monitoredRun(Task.java:134)
	at ghidra.util.task.TaskRunner.lambda$startTaskThread$0(TaskRunner.java:106)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: ghidra.program.model.mem.MemoryAccessException: Couldn't get all bytes for CodeUnit
	at ghidra.program.database.code.CodeUnitDB.getBytes(CodeUnitDB.java:653)
	at ghidra.app.plugin.match.MatchData.extractSourceHashes(MatchData.java:80)
	... 11 more

---------------------------------------------------
Build Date: 2024-Jan-30 1212 EST
Ghidra Version: 11.0.1
Java Home: <...>
JVM Version: Eclipse Adoptium 17.0.10
OS: Windows 11 10.0 amd64
```

**Expected behavior**
Generally, a successful no-exception automatic version tracking completion would be expected regardless of any match status. 

**Screenshots**
N/A, see steps/exception above.

**Attachments**
N/A, see steps/exception above.

**Environment (please complete the following information):**
 - OS: Windows 11 22H2
 - Java Version: Eclipse Adoptium 17.0.10
 - Ghidra Version: 11.0.1
 - Ghidra Origin: official GitHub distro

**Additional context**
This issue was encountered when trying to create a simple example for yet another Version Tracker bug causing a different exception. I hit the above instead, removed my test code for the other issue, and discovered this issue's exception with a just a simple "hello world" app (which pulls in RTL, C++ STL but relatively simple).

**Comment by AshleyT3** [comment]
It seems DataMatchProgramCorrelator is not handling uninitialized memory properly as does other parts of Ghidra. The DataMatchProgramCorrelator fails on the same exception that CodeBrowser ignores when a user is scroll across the same memory. DataMatchProgramCorrelator should view two equal uninitialized areas as equal.

Details...

- At the exception stack of the original post of this issue (above)...
- The initial address during the relevant getBytes call is Address.offset == 5368828384 which is 0x14001D1E0.
- The getBytes request needs a total of 152 bytes.
- The MemoryBlockDB instances (as assessed during MemoryMapDB.getBlockDB)... 
```
tmpBlocks	(id=449)	
    [0]	MemoryBlockDB  (id=571)	
    [1]	MemoryBlockDB  (id=572)	
    [2]	MemoryBlockDB  (id=439)	
    [3]	MemoryBlockDB  (id=573)	
    [4]	MemoryBlockDB  (id=379)	
        adapter	MemoryMapDBAdapterV3  (id=441)	
        id	4	
        invalid	false	
        lastSubBlock	FileBytesSubMemoryBlock  (id=386)	
        length	512	
        mappedBlocks	null	
        memMap	MemoryMapDB  (id=372)	
        record	DBRecord  (id=583)	
        startAddress	GenericAddress  (id=585)	
        subBlocks	ArrayList<E>  (id=451)	
            [0]	FileBytesSubMemoryBlock  (id=386)	// 32 bytes satisfied here, ok.
    [5]	MemoryBlockDB  (id=389)	
        adapter	MemoryMapDBAdapterV3  (id=441)	
        id	5	
        invalid	false	
        lastSubBlock	UninitializedSubMemoryBlock  (id=460)	
        length	1856	
        mappedBlocks	null	
        memMap	MemoryMapDB  (id=372)	
        record	DBRecord  (id=593)	
        startAddress	GenericAddress  (id=594)	
        subBlocks	ArrayList<E>  (id=481)	
            [0]	UninitializedSubMemoryBlock  (id=460)   // uninitialized, cannot satisfy the remaining 120 bytes.
```
- The getBytes request satisfies 32 bytes at the MemoryBlockDB of index 4.
- It fails trying get the remaining 120 bytes from the UninitializedSubMemoryBlock at index 5.
- Following is what CodeBrowser shows at the start address of the failing getBytes request. There are 32 valid bytes at 14001d1e0 followed by a gap and uninitialized memory.
```
14001d1e0 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 _EXCEPTION_RECORD
          00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
14001d278 ??
14001d279 ??
```

**Comment by ghidra007** [comment]
As you have mentioned in your follow on, it looks like the Data correlator is the culprit whether or not called from Auto VT. Thanks for tracking down the underlying issue.

**Comment by AshleyT3** [comment]
> Would you mind attaching the two files so we can see exactly what you are working with? It is not easy to get the same exact binary without knowing all of your compile settings.

I prefer to avoid sharing physical binaries publicly online if reasonable. In this case, it's tremendously easy to recreate the binaries using Visual Studio 2022 Community Edition with its default Windows C++ Console App project. Would you be able to do that? (See steps in prior post and let me know.)

**Comment by ghidra007** [comment]
> > Would you mind attaching the two files so we can see exactly what you are working with? It is not easy to get the same exact binary without knowing all of your compile settings.
> 
> I prefer to avoid sharing physical binaries publicly online if reasonable. In this case, it's tremendously easy to recreate the binaries using Visual Studio 2022 Community Edition with its default Windows C++ Console App project. Would you be able to do that? (See steps in prior post and let me know.)

No worries. We will do so and hopefully our setup will build the exact same ones or similar enough ones. Thanks.

**Comment by AshleyT3** [comment]
I dove in a little further after yesterday's update... here are some additional details fwiw...

## Summary
An affected PE includes both `.data` and `.bss` within the physical PE `.data` section, where portions of `.bss` (uninit data) can reside in the initialized portion of the `.data` section. Portions of `.bss` beyond the physical `.data` section size (virtualSize-rawDataSize in details below) is deemed "uninitialized" by Ghidra, yet those blocks may not include all of the logically uninitialized data, a byproduct of physical PE construction and Ghidra's section-to-block processing.

## Details...

Upon .exe import, `ghidra.app.util.opinion.PeLoader.processMemoryBlocks(...)` is called...
```
Daemon Thread [Task - Import File] (Suspended)	
	PeLoader.processMemoryBlocks(PortableExecutable, Program...) line: 673	
	PeLoader.load(...) line: 125	
	PeLoader(...).doLoad(...) line: 883	
	PeLoader(...).loadProgram(...) line: 98	
	PeLoader(...).load(...) line: 131	
	ImporterUtilities.importSingleFile(...) line: 395	
...
```

The `processMemoryBlocks` function processes each PE section, using virtualSize and rawDataSize to guide creation of initialized/uninitialized memory blocks.

For example, a `.text` section might have the following values...
  ```
  sectionName ... ".text" (id=669)
  addr .......... 5368713216 [0x140001000]
  virtualSize ... 4608 [0x1200]
  rawDataSize ... 4608 [0x1200]
  ```

The rawDataSize is deemed initialized data contained within the PE's section itself.

Any amount of virtualSize that is above rawDataSize is deemed uninitialized data.

Therefore, for each PE section where rawDataSize >0 bytes, `processMemoryBlocks` calls `MemoryBlockUtils.createInitializedBlock` to create an initialized data memory block.

For the above `.text` section example, since `rawDataSize==virtualSize`  (4608==4608), no uninitialized block is created for the `.text` section.

In the case reproducing the issue, the PE's physical `.data` section includes both the logical `.data` and `.bss` (uninitialized data) sections. I coin "logical" here to emphasize that `.bss` is a section to the developer concerned with link-time behaviors/outcomes so `.bss` exists as a section for uninitialized data in that sense, yet such is not reflected physically in the PE causing the issue to arise.

From the same example PE reproducing the issue, here are the values for its `.data` section...
  ```
  sectionName ... ".data" (id=696)  // includes .bss
  addr .......... 5368725504 [0x140004000]
  virtualSize ... 1011 [0x3f3]      // larger than area in PE
  rawDataSize ... 512 [0x200]       // area in PE
  ```

Note that virtualSize is larger than rawDataSize.

Given the above, `processMemoryBlocks` calculates 499 bytes for uninitialized block size...
```
rawDataSize == 512 == initialized block size (includes portions of .bss data).
(virtualSize-rawDataSize) == Ghidra uninitialized block size.
1011 - 512 == 499 bytes for Ghidra uninitialized block size.
```

For the above `.data` section example, `processMemoryBlocks` calls `MemoryBlockUtils.createInitializedBlock` to create a 512 byte initialized data segment (which technically includes a portion of uninitialized data), and it follows up with a call to `MemoryBlockUtils.createUninitializedBlock` to create a 499 byte uninitialized block.

Perhaps a helpful way to view all this... there is no penalty beyond impact to the binary size by actually initializing uninitialized data with zeros. Therefore, there is no penalty (beyond binary size) for the linker to include a portion of .bss in the initialized section, which is what happens in the example.

So the `.data` (initialized data) section can include a portion of uninitialized data which is, in fact, initialized to zero. The implications to Ghidra's current section/block processing are that uninitialized data items may span what Ghidra views as both initialized and uninitialized blocks.

Given the above example, the following is derived from Ghidra's disassembly. Notice .data begins at 0x140004000 and 0x200 bytes are included in the PE, yet within that 0x200 is a portion of .bss (uninit data)...

```
                //
                // .data 
                // ram:140004000-ram:1400041ff
                //
                .data
                normal_data1
140004000  01                   uchar  01h
                normal_data2
140004001  02                   uchar  02h
                normal_data3
140004002  03                   uchar  03h
140004003  00                   ??     00h
...
14000410f  00                   ??     00h
                        .bss
                        my_bss
140004110  00  00  00  00  00  00  00  00  00      MYSTRUCT
           00  00  00  00  00  00  00  00  00 
           00  00  00  00  00  00  00  00  00...
140004110  00   uchar     '\0'  field1
140004111  00   uchar     '\0'  field2
140004112  00   uchar     '\0'  field3
140004113  00   uchar     '\0'  field4
140004114  00   ??        00h   field4_0x4
140004115  00   ??        00h   field5_0x5
140004116  00   ??        00h   field6_0x6
...
1400041fd  00   ??        00h   field237_0xed
1400041fe  00   ??        00h   field238_0xee
1400041ff  00   ??        00h   field239_0xef
140004200       ??        ??    field240_0xf0
140004201       ??        ??    field241_0xf1
140004202       ??        ??    field242_0xf2
...
1400042f8       ??        ??    field488_0x1
1400042f9       ??        ??    field489_0x1
1400042fa       ??        ??    field490_0x1
1400042fb       ??        ??    field491_0x1
1400042fc       ??        ??    field492_0x1
1400042fd       ??        ??    field493_0x1
1400042fe       ??        ??    field494_0x1
1400042ff       ??        ??    field495_0x1
140004300       uint64    ??    array
140004308       ??        ??    field497_0x1
140004309       ??        ??    field498_0x1
14000430a       ??        ??    field499_0x1
...
140004375       ??        ??    field606_0x2
140004376       ??        ??    field607_0x2
140004377       ??        ??    field608_0x2

```

Relating to the details of my previous post, during Automatic Version Tracking, the VT correlation ends up calling into `ghidra.program.database.mem.MemoryMapDB.getBytes(...)` ...
```
Daemon Thread [Task - Auto Version Tracking Command] (Suspended)	
    MemoryMapDB.getBytes(Address, byte[], int, int, boolean) line: 1439
    MemoryMapDB.getBytes(Address, byte[], boolean) line: 1432
    DataDB(CodeUnitDB).getBytes() line: 652
    MatchData.extractSourceHashes(Program, AddressSetView, ...) line: 80
    MatchData.matchData(Program, AddressSetView, Program, AddressSetView, ...) line: 47
    DataMatchProgramCorrelator.doCorrelate(VTMatchSet, ...) line: 61
    ...
```

The `getBytes` function sees the initialized block, is able to get the bytes, but hits the uninitialized block on the second iteration to complete the `getBytes` request.

For the heck of it, I tried out the following quick test which allows auto VT to complete by allowing the caller of `getBytes` (the correlator in this case) to request that uninitialized blocks be treated as zero bytes...
```
public int getBytes(Address addr, byte[] dest, int dIndex, int size, boolean zeroBytesForUninitialized)
        throws MemoryAccessException {
    ...
    int numRead = 0;
    long lastRead = 0;
    while (numRead < size) {
        try {
            addr = addr.addNoWrap(lastRead);
            MemoryBlock block = getBlock(addr);
            ...
            int destOffset = numRead + dIndex;
            int numToRead = size - numRead;
            if (block.isInitialized() || block.isMapped()) {
                lastRead = block.getBytes(addr, dest, destOffset, numToRead);
            }
            else if (!block.isInitialized() && zeroBytesForUninitialized) {
                Arrays.fill(dest, destOffset, destOffset + numToRead, (byte)0);
                lastRead = numToRead;
            }
            else {
                break;
            }
            numRead += lastRead;
        }
        ...
    }
    ...
    return numRead;		
}
```
One downside to assuming zero bytes for uninitialized data would be a case where a previously uninitialized area is later modified to be initialized to zeros which would then hide the change itself from the correlator making the "zero-byte" assumption. It seems the best for correlation to consider uninit data to be only equal to contextually matching uninit data to mitigate a correlator from missing changes relevant to what it should detect (i.e., a prev unint area changed to an init area of same size should be a detected change, it seems).

**Comment by AshleyT3** [comment]
> > > Would you mind attaching the two files so we can see exactly what you are working with? It is not easy to get the same exact binary without knowing all of your compile settings.
> > 
> > 
> > I prefer to avoid sharing physical binaries publicly online if reasonable. In this case, it's tremendously easy to recreate the binaries using Visual Studio 2022 Community Edition with its default Windows C++ Console App project. Would you be able to do that? (See steps in prior post and let me know.)
> 
> No worries. We will do so and hopefully our setup will build the exact same ones or similar enough ones. Thanks.

With the current build tools, Visual Studio 2022 Community Edition, it is extraordinarily easy to reproduce this issue.

I would be surprised if you found a way to build a simple .exe using typical steps that would avoid the issue. I say this because of what's mentioned above... so long as you create an .exe that has a .data section whose binary data size is smaller than its virtual size, you should hit this issue. Generally, I see .data sections with 0x200 bytes in the PE and a virtual size larger than that 0x200 bytes... this is what causes the init/uninit blocks to be created, if that makes sense.

...but definitely let me know if anything is not easy to reproduce.

**Comment by AshleyT3** [comment]
If you want something more minimal, the following may be helpful/interesting...

You can use a .asm and Visual Studio 2022's ml64.exe to build a minimal .exe that may help remove noise and see a simpler layout. The example PE sections of my previous post were from a PE created with the following steps. 

## Steps:
1. Install/start [Visual Studio 2022 Community Edition](https://visualstudio.microsoft.com/vs/community/) on Windows 11. (Windows 10 is likely fine too.)
2. Press Start Menu ("Super Key" in Linux lingo), type "2022" to show Visual Studio 2022 command line shortcuts, click/select "x64 Native Tools Command Prompt for VS 2022" to open an x64 tools prompt (aka x64 native tools terminal window). 
3. Optionally, within the new terminal window, type "powershell" and press ENTER if you prefer PowerShell.
4. Create a new .asm source text file, Simple.asm, and copy/paste the following source code into that file...
```
; Build: ml64 /Fl /Zi .\Simple.asm /link /subsystem:console /entry:main /DEBUG

CONST SEGMENT PARA 'const'
sz_howdy_world db "Howdy there, world!",0
CONST ENDS

MYSTRUCT STRUCT 16
    field1 db ?
    field2 db ?
    field3 db ?
    field4 db 490 DUP(?)
    array dq 15 DUP(?)
MYSTRUCT ENDS

.DATA
normal_data1 db 1
normal_data2 db 2
normal_data3 db 3

.DATA?
my_bss MYSTRUCT <>

.CODE
main PROC
    sub rsp,28h                                     ; Allocate shadow space
    mov rax,offset normal_data1
    mov rax,offset normal_data2
    mov rax,offset normal_data3
    mov rax,offset my_bss
    mov BYTE PTR [rax],0aah
    add rsp,28h  ; Deallocate shadow space
    xor eax,eax                                     ; exit code is 0 success.
    ret
main ENDP
END
```
5. Run `ml64 /Fl /Zi .\Simple.asm /link /subsystem:console /entry:main /DEBUG` to assemble/link to PE .exe. In the same directory, you will now see Simple.exe and you may want to gander at Simple.lst, the assembly listing file.

The remainder of the steps are essentially the same as those in my original post...

6. Copy Simple.exe to Simple2.exe or some other name to create to versions. Differences are not important, they can match 100% and the issue is still observable. Alternatively, copy Simple.asm to Simple2.asm and build that to create the other binary. Goal is obviously to have two distinct physical binaries.
7. Start Ghidra.
8. File/New Project, Non-Shared Project, Project Name is SomeVtProj, click Finish.
9. Press I to import, locate and select Simple.exe, click Select File To Import, click OK on the next Import dialog, click OK on the Import Results Summary dialog.
10. Repeat step 9 but for Simple2.exe
11. Open Simple.exe in the CodeBrowser, from the CodeBrowser's menu choose Edit/Symbol Server Config, click +, Directory, specify the path to the Simple.exe pdb file, click OK, click OK. (It is assumed Simple.pdb and Simple2.pdb are in the same directory.) 
12. From the same CodeBrowser menu, use File/Load PDB File..., click Load, it should load Simple.pdb, click OK to dismiss warnings.
13. Close that CodeBrowser.
14. At this point, you can open each of Simple.exe and Simple2.exe if you wish to perform analysis ahead of time instead of when Version Tracking opens the CodeBrowser windows. Either way, ensure analysis is complete prior to actually starting Automatic Version Tracking.
15. Highlight both Simple.exe and Simple2.exe, right-click and choose Open With, Version Tracking, Simple.exe is the Source Program, Simple2.exe is the Destination Program, click Next, click Run Precondition Checks, click Next, click Finish...
16. Choose to perform any auto-analysis if it was not already done. Do this for both Simple.exe and Simple2.exe.
17. Go to Version Tracking Window, File/Automatic Version Tracking, observe the following exception...
ghidra.program.model.mem.MemoryAccessException: Couldn't get all bytes for CodeUnit. (See original post for full stack.)

If interested, from the same Visual Studio x64 native tools terminal window, you can use dumpbin to see the section details.

For example, the following shows the `.text` section as having a raw size larger than its virtual size, whereas the physical `.data` section (inclusive of `.bss`) has a raw size of only 0x200 bytes while its actual (virtual) size is 0x3F3 bytes. The uninitialized `my_bss` struct of the above .asm file spans from the PE's 0x200 initialized area to what will eventually become (upon loading) the uninitialized area of the overall physical .data section...

```
dumpbin /HEADERS Simple.exe
...
SECTION HEADER #1
   .text name
    1057 virtual size
    1000 virtual address (0000000140001000 to 0000000140002056)
    1200 size of raw data
     400 file pointer to raw data (00000400 to 000015FF)
       0 file pointer to relocation table
       0 file pointer to line numbers
       0 number of relocations
       0 number of line numbers
60000020 flags
         Code
         Execute Read
...
SECTION HEADER #3
   .data name
     3F3 virtual size
    4000 virtual address (0000000140004000 to 00000001400043F2)
     200 size of raw data
    1A00 file pointer to raw data (00001A00 to 00001BFF)
       0 file pointer to relocation table
       0 file pointer to line numbers
       0 number of relocations
       0 number of line numbers
C0000040 flags
         Initialized Data
         Read Write
...
```

**Comment by AshleyT3** [comment]
Okay, likely not needed, but if you are interested, here is a very simple minimal example that will pass or fail Automatic Version Tracking (AVT) depending on whether you build it with /DFAILING_CASE or not. The difference is 1 byte of initialized data that, if present, causes the first `.bss` variable to start at the beginning of the `.data` section beyond the raw data included within the PE.

```
; Build for success case:
;   ml64.exe /Zi .\SimplePassFail.asm /link /subsystem:console /entry:main /DEBUG
; Build for failing case:
;   ml64.exe /DFAILING_CASE /Zi .\SimplePassFail.asm /link /subsystem:console /entry:main /DEBUG

IFDEF FAILING_CASE
NUM_FILL_BYTES EQU 200
ELSE
NUM_FILL_BYTES EQU 201
ENDIF

MYSTRUCT STRUCT 16
    field1 db ?
    field2 db ?
    field3 db ?
    field4 db 490 DUP(?)
    array dq 15 DUP(?)
MYSTRUCT ENDS

.DATA
fill_var db NUM_FILL_BYTES DUP(0)

.DATA?
my_bss MYSTRUCT <>

.CODE
main PROC
    sub rsp,28h                                     ; Allocate shadow space
    mov rax,offset my_bss
    mov BYTE PTR [rax],0aah
    mov rax,offset fill_var
    mov BYTE PTR [rax],0bbh
    add rsp,28h  ; Deallocate shadow space
    xor eax,eax                                     ; exit code is 0 success.
    ret
main ENDP
END
```

**Comment by ryanmkurtz** [comment]
Thank you for the detailed instructions...I was able to follow your original post and reproduce the issue with VS2022.  All of our loaders create uninitialized blocks for memory that is not backed by the original file.  This is done for reasons of efficiency....some of them can be quite large.  This is not how real memory works of course...there is no concept of "uninitialized" there.  But, we also don't want to precisely model all of memory...or we'd need a lot of memory.  So, as you've noticed, all of our code that works with memory needs to account for both initialized and uninitialized blocks.

**Comment by AshleyT3** [comment]
> Thank you for the detailed instructions...I was able to follow your original post and reproduce the issue with VS2022. All of our loaders create uninitialized blocks for memory that is not backed by the original file. This is done for reasons of efficiency....some of them can be quite large. This is not how real memory works of course...there is no concept of "uninitialized" there. But, we also don't want to precisely model all of memory...or we'd need a lot of memory. So, as you've noticed, all of our code that works with memory needs to account for both initialized and uninitialized blocks.

Thank you @ryanmkurtz (and @ghidra007). What you're saying makes great sense.

If it helps temper prioritization, this issue is not pressing for myself personally. I was investigating what Ghidra had available in this area but am not blocked by this issue, and I do have the aforementioned workaround as needed.

An aside, not shirking here... I'd have pushed out a first Ghidra PR had path been clear but there seemed many potential avenues each affecting internal areas in nuanced ways... seemed best to report, bounce off you.
